spring-web-reactive/src/main/java/org/springframework/web/reactive/result/method/annotation/ServerHttpMessageReader.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2002-2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.web.reactive.result.method.annotation;
```

Comment:
`ServerHttpMessageWriter` is generally useful for sever-side use. As such I see it as belonging in spring-web next to `HttpMessageWriter` where we also have other server-side only writers like the one for Server-Sent Events. 

The `ResourceHttpMessageWriter` needs it and that's another confirmation it should be there. It would be needed for JSONP as well and anything else that needs access to the request to prepare hints.

As for the `MethodParameter` couldn't we make that a `ResolvableType`, call it `streamType`, and order it ahead of `elementType` in the method signature? That would be a little less assuming that there is a `MethodParameter` necessarily.